src/pymor/algorithms/symplectic.py

```
@@ -0,0 +1,474 @@
+        return result
+
+    def extend(self, U, method='svd_like', modes=2, product=None):
+        """Extend the |SymplecticBasis| with vectors from a |VectorArray|.
+
+        Parameters
+        ----------
+        U
+            The vectors used for the extension as |VectorArray|.
+        method
+            The method used for extension. Available options are
+                ('svd_like', 'complex_svd', 'symplectic_gram_schmidt').
+        modes
+            Number of modes to extract from U. Has to be even.
+        product
+            A product to use for the projection error. Default is None.
+        """
+        from pymor.algorithms.symplectic import (psd_complex_svd,
+                                                 psd_svd_like_decomp,
+                                                 symplectic_gram_schmidt)
+        assert modes % 2 == 0, 'number of modes has to be even'
+        assert method in ('svd_like', 'complex_svd', 'symplectic_gram_schmidt')
+
+        U_proj_err = U - self.lincomb(U.inner(self.transposed_symplectic_inverse().to_array()))
+        proj_error = U_proj_err.norm(product=product)
+
+        if method in ('svd_like', 'complex_svd'):
+
+            if method == 'svd_like':
+                new_basis = psd_svd_like_decomp(U_proj_err, self.J, modes)
+            elif method == 'complex_svd':
+                new_basis = psd_complex_svd(U_proj_err, self.J, modes)
+
+            self.append(new_basis)
+
+        elif method == 'symplectic_gram_schmidt':
+            basis_length = len(self)
+            # find max error
+            idx = proj_error.argsort()[-modes//2:][::-1]
+            new_basis = U[idx].copy()
+            new_basis.scal(1/new_basis.norm())
+            new_basis.append(self.J.apply_adjoint(new_basis))
+            self.append(SymplecticBasis.from_array(new_basis, self.J))
+            symplectic_gram_schmidt(self.E, self.F, self.J, offset=basis_length, copy=False)
```

Do you think the `extend` should raise an error if an unknown method is passed? Currently it just wouldn't do anything.

---

`method in ('svd_like', 'complex_svd', 'symplectic_gram_schmidt')` is checked in line 180, but for a clearer code, I can add an else-case, sure

---

Sorry, I did not see that line. No need to change anything.